python-package/xgboost/core.py

```diff
@@ -434,14 +437,26 @@ class DMatrix:                  # pylint: disable=too-many-instance-attributes
     which is optimized for both memory efficiency and training speed.
     You can construct DMatrix from multiple different sources of data.
     """
-
-    def __init__(self, data, label=None, weight=None, base_margin=None,
-                 missing=None,
-                 silent=False,
-                 feature_names=None,
-                 feature_types=None,
-                 nthread=None,
-                 enable_categorical=False):
+    @_deprecate_positional_args
+    def __init__(
+        self,
+        data,
+        label=None,
+        *,
+        weight=None,
+        base_margin=None,
+        group=None,
+        qid=None,
+        label_lower_bound=None,
+        label_upper_bound=None,
+        missing: Optional[float] = None,
```

Comment:
Considering the number of meta info is growing, I think it's reasonable choice to stick with positional only arguments.

Reply:
> it's reasonable choice to stick with positional only arguments.

Did you mean to say "keyword-only arguments"?

Reply:
Yup, sorry.

Reply:
It is possible to delay the breaking change by moving the `missing` parameter to right after `base_margin` so as to preserve the current parameter order. This would give users some time to switch over to keyword-only args.

Is there a good reason for moving the `missing` parameter right now?

Reply:
I will try to move parameters.

Reply:
If the re-ordering is to make Dask and non-Dask interface consistent, then I'm okay with the breaking change. Is this the case here?

Reply:
I think you are right that it doesn't have to be breaking change.  Let me try a bit harder after the fixing up the mess I made with the patch rel.